### ISSUE TITLE
Cannot cast T to unknown

### ISSUE THREAD

**Issue by szmarczak** [microsoft/TypeScript#39429]
**Search Terms:** Cannot cast T to unknown

**Code**

```ts
interface Options<T> {
    fn: (value: T) => void
}

const normalizeOptions = (options: Options<unknown>): void => {};

function someOtherFunction<T>(options: Options<T>) {
    normalizeOptions(options); // The `options` has a red underline
}
```

**Expected behavior:** No error.

**Actual behavior:**

```
Argument of type 'Options<T>' is not assignable to parameter of type 'Options<unknown>'.
  Type 'unknown' is not assignable to type 'T'.
    'T' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.(2345)
```

**Playground Link:** https://www.typescriptlang.org/play?ts=3.9.2#code/JYOwLgpgTgZghgYwgAgPIAczAPYgM4A8AKgHzIDeAUMjcjCAFzIAUAbnADYCuETRAlMgC8ZVtmAATSgF9KlBLjxhkIbFAC2nYAC8IGLIuEtsmHPib6zhLiADWqgO4gS-JmMnCy5aQG45MGwQDEGQ8bHU9MAALaAAxQODiEmYTYLwLU0UkwSpaFTVNDh09TPwU0rx+P1kgA

**Related Issues:** I don't know.

/cc @sindresorhus

**Comment by jack-williams** [comment]
`Options` is contravariant in `T` because `T` appears in a function argument position, so it is the expected behaviour not to be able to assign `Options<T>` to `Options<unknown`.

If you're happy to take the unsoundness and want to turn off the error, you can declare `Options` using a (bivariant) method, rather than a function property.

```ts
interface Options<T> {
    fn(value: T): void
}
```

**Comment by szmarczak** [comment]
@jack-williams I don't think I'll go with your approach. https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/method-signature-style.md

> T appears in a function argument position, so it is the expected behaviour not to be able to assign Options<T> to Options<unknown.

I disagree.

> If you're happy to take the unsoundness and want to turn off the error,

I want to, but this is a simple case. What about 

```ts
interface Options<T> {
    fns: ((value: T) => void)[]
}

const normalizeOptions = (options: Options<unknown>): void => {};

function someOtherFunction<T>(options: Options<T>) {
    normalizeOptions(options);
}
```

**Comment by AlCalzone** [comment]
@szmarczak This seems to work:
```ts
interface Options<T> {
    fn: (value: T) => void
}

const normalizeOptions = <T extends unknown>(options: Options<T>): void => {};

function someOtherFunction<T>(options: Options<T>) {
    normalizeOptions(options);
}
```

**Comment by szmarczak** [comment]
@AlCalzone That works. But isn't `(options: Options<unknown>): void` and `<T extends unknown>(options: Options<T>): void` the same?

**Comment by szmarczak** [comment]
Ok I got it. `Options<unknown>` behaves like `Options<any>` in this case. You could do `options.fn('asdf');` and it would work since `string` can be casted to `unknown`. `T extends unknown` says that T can be anything, a string, a number etc. hence `options.fn('asdf');` will fail.

**Comment by szmarczak** [comment]
@jack-williams @AlCalzone Thanks for explaining! I really appreciate it.

**Comment by ThaJay** [comment]
@AlCalzone 
`Constraining the generic type 'T' to 'unknown' does nothing and is unnecessary.`

This feels so stupid I would have expected `unknown` to accept all types because the word tells me it is something but we don't know what exactly.

> @szmarczak This seems to work:
> 
> ```ts
> interface Options<T> {
>     fn: (value: T) => void
> }
> 
> const normalizeOptions = <T extends unknown>(options: Options<T>): void => {};
> 
> function someOtherFunction<T>(options: Options<T>) {
>     normalizeOptions(options);
> }
> ```
